feat: add property to events + policy updates

httpJunkie opened the pull request:
## Description

This PR adds a property to the 'Transaction Finalized' events. This change:

- Implements support for a remote feature flag
- Adds proper type support for remote feature flags
- Ensures compatibility with both regular and Smart Transactions

## Related issues

[Part of: #4375](https://github.com/MetaMask/MetaMask-planning/issues/4375)

See Notion Document for full PR details:
[Notion Document](https://www.notion.so/metamask-consensys/TX-Hash-in-Analytics-and-Privacy-Policy-Toast-Update-Extension-1b9f86d67d6880e9809ef7229f0475d0?d=1c2f86d67d6880138761001cfa64f53a#1c0f86d67d6880369fbbc748ca194261)

## Manual testing steps

1. Enable MetaMetrics in your wallet settings
2. Perform tests in Testing Plan 

*Refer to [Notion Document](https://www.notion.so/metamask-consensys/TX-Hash-in-Analytics-and-Privacy-Policy-Toast-Update-Extension-1b9f86d67d6880e9809ef7229f0475d0?d=1c2f86d67d6880138761001cfa64f53a#1c0f86d67d6880369fbbc748ca194261) for:

- Details on this PR and it's changes
- Testing Plan
- Screenshots showing functionality working as of latest commit to this branch

## **Screenshots/Recordings**

See: "PROOF WITH FLAGS ENABLED" & "PROOF WITH FLAGS DISABLED" in:
[Notion Document](https://www.notion.so/metamask-consensys/TX-Hash-in-Analytics-and-Privacy-Policy-Toast-Update-Extension-1b9f86d67d6880e9809ef7229f0475d0?d=1c2f86d67d6880138761001cfa64f53a#1c0f86d67d6880369fbbc748ca194261)

### **Before**

See: "PROOF WITH FLAGS DISABLED" in:
[Notion Document](https://www.notion.so/metamask-consensys/TX-Hash-in-Analytics-and-Privacy-Policy-Toast-Update-Extension-1b9f86d67d6880e9809ef7229f0475d0?d=1c2f86d67d6880138761001cfa64f53a#1c0f86d67d6880369fbbc748ca194261)

### **After**

See: "PROOF WITH FLAGS ENABLED" in:
[Notion Document](https://www.notion.so/metamask-consensys/TX-Hash-in-Analytics-and-Privacy-Policy-Toast-Update-Extension-1b9f86d67d6880e9809ef7229f0475d0?d=1c2f86d67d6880138761001cfa64f53a#1c0f86d67d6880369fbbc748ca194261)

## **Pre-merge author checklist**

- [x] I've followed [MetaMask Contributor Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask Extension Coding Standards](https://github.com/MetaMask/metamask-extension/blob/main/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've completed the PR template to the best of my ability
- [x] I’ve included tests if applicable
- [x] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [x] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-extension/blob/main/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [x] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [x] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

httpJunkie committed:
feat: Add transaction_hash to MetaMetrics events

Add transaction hash property to "Transaction Finalized" events for users
who opt-in to analytics. Implement feature flag integration for controlled
rollout via Launch Darkly. Update types and tests to support the feature.

httpJunkie committed:
test: add `getRemoteFeatureFlags` mock to transactions metrics test

httpJunkie committed:
fix: Add transaction_hash type to CrossChainSwapsEventProperties

Update type definitions to properly include `transaction_hash` in `CrossChainSwapsEventProperties`,
resolving TypeScript errors. Replace any type assertions with strongly typed properties.

httpJunkie committed:
feat: Use remote feature flags for PP Toast

Replace hardcoded pp date with LD feature flag to control toast visibility. Add event tracking for user interactions with
the toast. This implementation allows for smoother rollout and reduces
effort for future pp updates.

httpJunkie committed:
feat: update to selectors to try and fix the date comparison issue.

httpJunkie committed:
Fix Privacy Policy Toast functionality and onboarding date handling

- Fix toast dismissal by adding null checks for platform.trackEvent
- Improve error handling in setNewPrivacyPolicyToastClickedOrClosed
- Refactor selectShowPrivacyPolicyToast for better date comparison logic
- Only set onboardingDate if not already set and for new wallets
- Add debug logging for better troubleshooting
- Remove unused imports and console logs

httpJunkie committed:
Merge branch 'main' into feat/tx-finalized-add-hash-and-toast

httpJunkie committed:
feat: remove logging

httpJunkie committed:
fix: uncomment toast we were having issues with but not related

httpJunkie committed:
fix: add missing NFT detection toast selector

The `selectNftDetectionEnablementToast` selector was being imported but not exported
from the selectors file, causing a compilation error. This commit adds the missing
selector to properly handle the NFT detection enablement toast state.

The selector returns a boolean from `state.appState.showNftDetectionEnablementToast`
which is used by the `ToastMaster` component to conditionally render the NFT
detection toast.

httpJunkie committed:
fix: update linting issues with prettier problems

httpJunkie committed:
Merge branch 'main' into feat/tx-finalized-add-hash-and-toast

httpJunkie committed:
feat: Implement transaction hash in metrics and privacy policy toast

- Add transaction_hash to "Transaction Finalized" events for MetaMetrics opt-in users
- Add getRemoteFeatureFlags method to TransactionMetricsRequest
- Modify transaction metrics code to check Launch Darkly flag before adding hash
- Fix onboardingDate settings to only set for new wallets
- Update privacy policy toast to use Launch Darkly feature flag
- Add proper error handling for privacy policy toast interactions
- Add tracking for toast "Read More" and "Close" actions

httpJunkie committed:
chore: remove logs from previous implementation

httpJunkie committed:
Merge branch 'main' into feat/tx-finalized-add-hash-and-toast

httpJunkie committed:
test: Add transaction hash in analytics and privacy policy toast

- Implement transaction hash addition to metrics for opt-in users
- Add transaction_hash to Transaction Finalized and Swap events
- Add getRemoteFeatureFlags method to TransactionMetricsRequest
- Update privacy policy toast to use Launch Darkly flag
- Fix onboardingDate logic to only set for new wallet creation
- Add proper error handling for toast interactions
- Add tracking for toast user actions
- Update tests to work with remote feature flags

This implements both the transaction hash analytics feature and
the privacy policy toast update using Launch Darkly feature flags
for controlled rollout.

httpJunkie committed:
Merge branch 'main' into feat/tx-finalized-add-hash-and-toast

httpJunkie commented:
I have read the CLA Document and I hereby sign the CLA

httpJunkie committed:
Merge branch 'main' into feat/tx-finalized-add-hash-and-toast

httpJunkie commented on app/scripts/lib/bridge-status/metrics.ts:
```diff
@@ -108,6 +115,14 @@ export const handleBridgeTransactionComplete = async (
 
+  if (
+    getRemoteFeatureFlags()['transactions-tx-hash-in-analytics'] &&
+    getParticipateInMetrics() &&
```
I really appreciate this suggestion and I'm looking into it further.

My initial response (even if I decide to remove it)

There may be a check in the `trackEvent` function, but IMHO having this check at the point of adding the transaction hash is important to me for several reasons:

- Ensures sensitive data isn't accidentally included
- Original requirements explicitly state `transaction-hash` should only be added for users who opt into MetaMetrics
- Intent is clear
- If general metrics are blocked later, ensures (specifically) the transaction hash isn't added to regular properties

httpJunkie replied:
I compromised. I do not like the fact that I made that change and had to dig to find out if the check was made elsewhere. I added a comment in both:

- `app/scripts/lib/bridge-status/metrics.ts` &
- `app/scripts/lib/transaction/metrics.ts`

```ts
  if (
    transactionMetricsRequest.getRemoteFeatureFlags?.()
      ?.transactionsTxHashInAnalytics &&
    transactionMeta?.hash
  ) {
    // Note: getParticipateInMetrics() check happens in metametrics-controller.ts#submitEvent
    properties.transaction_hash = transactionMeta.hash;
  }
```

I think this is better than being intentional and redundant. LMKWYT

httpJunkie commented on app/scripts/metamask-controller.js:
```diff
@@ -1713,6 +1713,11 @@ export default class MetamaskController extends EventEmitter {
         const state = this._getMetaMaskState();
         return getFeatureFlagsByChainId(state);
       },
+      getRemoteFeatureFlags: () => {
```
Good catch

httpJunkie replied:
Removed with comment on why it wasn't needed.

httpJunkie committed:
fix: remove commented import

httpJunkie commented on app/scripts/lib/transaction/metrics.ts:
```diff
@@ -1083,6 +1096,16 @@ async function buildEventFragmentProperties({
+  if (
+    transactionMetricsRequest.getRemoteFeatureFlags?.()
+      ?.transactionsTxHashInAnalytics &&
+    transactionMetricsRequest.getParticipateInMetrics() &&
```
While I'm checking on your previous comment... I just noticed that I'm accessing the same flag differently in each file. The difference is:

In `transaction/metrics.ts`, we're using dot notation and accessing the property as transactionsTxHashInAnalytics (camelCase)
In `bridge-status/metrics.ts`, we're using bracket notation and accessing the property as 'transactions-tx-hash-in-analytics' (kebab-case)

I'm actually pretty sure they both work but I want to be consistent, but I need to figure out what I'm going to do regarding your last comment and I need to test real quick.

httpJunkie committed:
fix: use dot notation for lib/bridge/metrics (wasn't working)

httpJunkie committed:
chore: Remove unused getRemoteFeatureFlags from SmartTransactionsController init

This method was added in anticipation of implementing transaction hash
analytics directly in the `SmartTransactionsController`, but the final
implementation handles this in the metrics utility files instead.

The transaction hash feature is properly managed through the existing
metrics pipeline without requiring direct feature flag access in the controller.

httpJunkie committed:
Merge branch 'main' into feat/tx-finalized-add-hash-and-toast

httpJunkie committed:
refactor: Remove redundant participateInMetrics checks in tx hash analytics

The getParticipateInMetrics() checks in transaction hash analytics
conditions are redundant since metametrics-controller.ts#submitEvent
already verifies user opt-in status before sending any metrics.

Added comments to document where this check happens to maintain code clarity
and ensure future developers understand the privacy safeguards in place.

httpJunkie committed:
refactor: Remove hardcoded privacy policy date in favor of Launch Darkly flag

Remove ui/helpers/constants/privacy-policy.ts file and update tests to use
remote feature flag for privacy policy updates. Tests now verify correct
toast behavior based on feature flag value rather than hardcoded date.

httpJunkie committed:
Merge branch 'main' into feat/tx-finalized-add-hash-and-toast

httpJunkie committed:
chore: update onboarding metrics message key

- Rename metrics collection message key for clarity
- Update related references in onboarding flow
- Remove deprecated message key from en/en-GB locales

httpJunkie committed:
fix: remove unused `getParticipateInMetrics` from metric.ts

httpJunkie committed:
test: update snapshots for metametrics

httpJunkie committed:
fix: run prettier in locale files (order)

httpJunkie committed:
Merge branch 'main' into feat/tx-finalized-add-hash-and-toast

httpJunkie committed:
fix: linting error

httpJunkie committed:
Merge branch 'main' into feat/tx-finalized-add-hash-and-toast

dan437 commented on app/scripts/controllers/app-state-controller.ts:
```diff
@@ -534,7 +534,10 @@ export class AppStateController extends BaseController<
-      state.onboardingDate = Date.now();
+      // Only set if not already set
+      if (!state.onboardingDate) {
+        state.onboardingDate = Date.now();
```
Nice!

dan437 commented on app/scripts/lib/transaction/metrics.ts:
```diff
@@ -351,6 +353,16 @@ export const handlePostTransactionBalanceUpdate = async (
   if (getParticipateInMetrics() && transactionMeta.swapMetaData) {
+    // Prepare the transaction hash addition if feature flag is enabled
+    const shouldAddTransactionHash =
+      getRemoteFeatureFlags &&
```
nit: there is a little shortcut: `getRemoteFeatureFlags?.()?.['transactions-tx-hash-in-analytics']`

dan437 commented on ui/components/app/toast-master/selectors.ts:
```diff
@@ -122,3 +138,7 @@ export function selectNewSrpAdded(state: State): boolean {
 ///: END:ONLY_INCLUDE_IF
+
+export function selectNftDetectionEnablementToast(state: State): boolean {
+  return Boolean(state.appState.showNftDetectionEnablementToast);
```
Don't we want to preserve code that was there before?
```return Boolean(state.appState?.showNftDetectionEnablementToast);```

httpJunkie committed:
Merge branch 'main' into feat/tx-finalized-add-hash-and-toast

httpJunkie committed:
Merge branch 'main' into feat/tx-finalized-add-hash-and-toast

httpJunkie committed:
Merge branch 'main' into feat/tx-finalized-add-hash-and-toast

httpJunkie committed:
feat: fix optional chaining in two places.

httpJunkie committed:
Merge branch 'main' into feat/tx-finalized-add-hash-and-toast

httpJunkie committed:
Merge branch 'main' into feat/tx-finalized-add-hash-and-toast

OGPoyraz commented on shared/types/metametrics.ts:
```diff
@@ -66,6 +66,10 @@ export type TransactionMetricsRequest = {
   getIsConfirmationAdvancedDetailsOpen: () => boolean;
   getHDEntropyIndex: () => number;
+  getRemoteFeatureFlags: () => {
+    'transactions-tx-hash-in-analytics'?: boolean;
```
I was wondering if you straight defined `transactions-tx-hash-in-analytics` in the LD root? Usually the approach is to collect commons under a group rather than root.

httpJunkie replied:
I'm going to mark as resolved, but if you know of something I'm missing we can review further. Feel free to check the Launch Darkly flags setup for Extension and Mobile and LMK if there is more I can do.

OGPoyraz reviewed (changes requested):
Reviewed the confirmation changes - I left couple comments, thanks for the PR!

davidmurdoch commented on ui/components/app/toast-master/selectors.ts:
```diff
@@ -62,25 +63,40 @@ export function selectShowPrivacyPolicyToast(state: State): {
+  }
+
+  // Create dates and ensure consistent timezone handling
+  const newPrivacyPolicyDate = new Date(policyUpdateDate);
```
`newPrivacyPolicyDate` should be validated, as a misconfiguration of this value could result in some strange behavior.

httpJunkie replied:
Absolutely. Making this change.

httpJunkie replied:
Added some validation for the feature flag's `newPrivacyPolicyDate`
Prevent showing toast if date is fat fingered or missing
Toast behavior should be consistent with valid  dates

httpJunkie committed:
fix: Standardize on camelCase for feature flag & ensure tx hash in regular properties

- Convert all kebab-case references to camelCase for consistency
- Remove transaction hash from sensitiveProperties (should only be in regular properties)
- Update type definitions to match implementation

httpJunkie commented on app/scripts/lib/bridge-status/metrics.ts:
```diff
@@ -108,6 +113,14 @@ export const handleBridgeTransactionComplete = async (
   };
 
+  if (
+    getRemoteFeatureFlags().transactionsTxHashInAnalytics &&
```
I've standardized all the usages to camelCase.

OGPoyraz replied:
We will get the literal key-value pairs from `getRemoteFeatureFlags` function so this has to be kebab case. (checking after LaunchDarkly)
![Screenshot 2025-04-09 at 08 43 40](https://github.com/user-attachments/assets/e33330d1-19c7-42c6-bcbd-ee1d65488ffb)

httpJunkie commented on app/scripts/lib/transaction/metrics.ts:
```diff
@@ -1093,6 +1105,15 @@ async function buildEventFragmentProperties({
+    transactionMeta?.hash
+  ) {
+    // Note: getParticipateInMetrics() check happens in metametrics-controller.ts#submitEvent
+    properties.transaction_hash = transactionMeta.hash;
```
Have removed from adding to sensitive properties. Originally I thought that is where it should be. They asked me to put in regular properties. It should only be in regular properties now.

OGPoyraz replied:
Does this warrant atleast a small unit test?

httpJunkie replied:
Added focused test to verify transaction hash is correctly added to metrics 
properties when the feature flag is enabled and transaction has a.... hash.

The test simply verifies only the functionality we added. I agree with the suggestion.

httpJunkie committed:
Merge branch 'main' into feat/tx-finalized-add-hash-and-toast

httpJunkie commented on app/scripts/controllers/app-state-controller.ts:
```diff
@@ -534,7 +534,10 @@ export class AppStateController extends BaseController<
     this.update((state) => {
-      state.onboardingDate = Date.now();
+      // Only set if not already set
+      if (!state.onboardingDate) {
+        state.onboardingDate = Date.now();
+      }
```
Hey, good catch — I updated the code to use `??=` ( logical nullish assignment operator) as you suggested.

It's cleaner and more precise. we only want to set onboarding date when it’s currently null or undefined, avoiding to unintentionally overwrite any existing value (like in the case of imported wallets). 

It also improves readbility "set-once-if-not-set" logic in a single line. I just rarely use it so thanks for the suggestion. Great catch. Would have worked without but I'm making changes so lets do it.

httpJunkie committed:
fix: improve onboarding date handling and privacy policy toast logic

- Only set onboarding date for new wallets to prevent incorrect toast behavior on imported wallets.
- Update onboarding date setting to ensure persistence for already set dates.
- Add validation to privacy policy update date to prevent issues from invalid or missing feature flag values.
- Refactor logic for showing the privacy policy toast based on valid date checks and onboarding status.

httpJunkie committed:
Merge branch 'main' into feat/tx-finalized-add-hash-and-toast

httpJunkie committed:
fix: remove `getRemoteFeatureFlags`

- Removed unused arguments and logic related to feature flags for transaction hash in analytics

httpJunkie committed:
Merge branch 'main' into feat/tx-finalized-add-hash-and-toast

httpJunkie committed:
Merge branch 'main' into feat/tx-finalized-add-hash-and-toast

httpJunkie committed:
Merge branch 'main' into feat/tx-finalized-add-hash-and-toast

davidmurdoch reviewed (changes requested):
leaving a blocking review as we discuss the scope of these changes

httpJunkie committed:
Merge branch 'main' into feat/tx-finalized-add-hash-and-toast

httpJunkie committed:
feat: update onboarding text and descriptions (phase 1)

httpJunkie committed:
feat: update onboarding message `onboardingMetametricsNeverSellDataWithLink` to make "settings" a link

httpJunkie committed:
test: Add test for transaction hash in metrics feature

Add focused test to verify transaction hash is correctly added to metrics
properties when the feature flag is enabled and transaction has a hash.

This test ensures our conditional logic works as expected without creating
brittle test cases that would fail on unrelated changes. The test verifies
only the specific functionality we added.

httpJunkie committed:
refactor: Remove redundant transaction hash feature flag check from bridge metrics

Removed duplicate feature flag check for adding transaction hashes to metrics
from bridge-status/metrics.ts, specifically in the handleBridgeTransactionComplete
function.

The transaction hash is already being added in the main buildEventFragmentProperties
function in transaction/metrics.ts which handles all transaction types including
bridge transactions. Extensive testing confirmed that:

1. The feature flag system is properly accessible in the transaction metrics
   context (getRemoteFeatureFlags returns valid data)
2. The transactionsTxHashInAnalytics flag is correctly detected when enabled
3. Transaction hashes are successfully added to metrics properties

This change simplifies our codebase by eliminating redundant code while
maintaining consistent behavior across all transaction types.

httpJunkie committed:
Merge branch 'main' into feat/tx-finalized-add-hash-and-toast

httpJunkie committed:
test: update snapshots for onboarding-flow

httpJunkie committed:
Merge branch 'main' into feat/tx-finalized-add-hash-and-toast

httpJunkie committed:
fix: run prettier on `app/_locales/en_GB/messages.json`
and  `app/_locales/en/messages.json`

httpJunkie committed:
Merge branch 'main' into feat/tx-finalized-add-hash-and-toast

httpJunkie committed:
fix: add `settingsLowercase` to en_GB

httpJunkie committed:
Merge branch 'main' into feat/tx-finalized-add-hash-and-toast

httpJunkie committed:
Merge branch 'main' into feat/tx-finalized-add-hash-and-toast

httpJunkie committed:
feat: update privacy policy toast text

httpJunkie committed:
fix: update privacy policy notice

httpJunkie committed:
Merge branch 'main' into feat/tx-finalized-add-hash-and-toast

httpJunkie committed:
fix: linting issues with locales

httpJunkie committed:
Merge branch 'main' into feat/tx-finalized-add-hash-and-toast

httpJunkie committed:
Merge branch 'main' into feat/tx-finalized-add-hash-and-toast

dan437 reviewed (dismissed):
Let's QA this carefully once the new release candidate is created.

httpJunkie commented on app/_locales/en/messages.json:
```diff
@@ -3304,8 +3304,8 @@
-  "newPrivacyPolicyTitle": {
-    "message": "We’ve updated our privacy policy"
+  "newPrivacyPolicyNotice": {
+    "message": "We’ve changed our metrics policy. We may collect data associated with public transactions on-chain. We never sell your data and you can opt out any time."
```
I can bring this up to the team, there are potentially a few text changes that could easily be done once the PR is merged. But this is the text they approved.

davidmurdoch replied:
I've brought it up in here https://consensys.slack.com/archives/C04N9GVC5DE/p1744296047780849?thread_ts=1744234085.460189&cid=C04N9GVC5DE

httpJunkie replied:
Ok, are we good to go with the review?
I will follow up on it.

davidmurdoch replied:
No, I never believe someone, including myself, when they say they'll fix a request in a future PR. :-)

httpJunkie replied:
Hopefully I can gain your trust.

DDDDDanica commented on ui/components/app/toast-master/selectors.ts:
```diff
@@ -62,25 +64,40 @@ export function selectShowPrivacyPolicyToast(state: State): {
-  const newPrivacyPolicyDate = new Date(PRIVACY_POLICY_DATE);
-  const currentDate = new Date(Date.now());
+
+  const remoteFeatureFlags = state.metamask?.remoteFeatureFlags || {};
```
You can use the `getRemoteFeatureFlags` selector from `ui/selectors/remote-feature-flags.ts`

httpJunkie replied:
I will revisit this. TY

DDDDDanica commented on app/scripts/metamask-controller.js:
```diff
@@ -7721,6 +7723,12 @@ export default class MetamaskController extends EventEmitter {
           .showConfirmationAdvancedDetails;
       },
       getHDEntropyIndex: this.getHDEntropyIndex.bind(this),
+      getRemoteFeatureFlags: () => {
```
If the reason is to access the state from UI, you can safely remove this function

httpJunkie replied:
I will revisit this. TY

DDDDDanica replied:
Can you create an ticket to track this? As this is not applying to current usage of remote feature flags and I want to make sure it's removed to avoid confusion for other developers to user remote feature flags

OGPoyraz reviewed (dismissed):
Confirmation changes LGTM

httpJunkie committed:
Merge branch 'main' into feat/tx-finalized-add-hash-and-toast

tayvano commented:
Contrary to the title of this pull request, the true purpose is to capture and collect transaction hashes for all users who have opted into metametrics.

This allows MetaMask to correlate all data collected in metametrics with a users onchain data which includes, but is not limited to, their activity onchain, across chains, and in-app.

This was not possible before and was explicitly disallowed by our privacy policy.

gnidan commented:
> This allows MetaMask to correlate all data collected in metametrics with a users onchain data which includes, but is not limited to, their activity onchain, across chains, and in-app.
> 
> This was not possible before and was explicitly disallowed by our privacy policy.

This sounds awful! Do people who opt-in realize that this correlation is going to happen?

Maybe it's time to divest. Does anyone want 16,000 shares of Consensys?

httpJunkie committed:
Merge branch 'main' into feat/tx-finalized-add-hash-and-toast

httpJunkie committed:
Merge branch 'main' into feat/tx-finalized-add-hash-and-toast

httpJunkie committed:
Merge branch 'main' into feat/tx-finalized-add-hash-and-toast
